src/librustdoc/doctest/make.rs

```diff
@@ -447,15 +419,17 @@ fn parse_source(
     let mut info =
         ParseSourceInfo { already_has_extern_crate: crate_name.is_none(), ..Default::default() };
 
+    const DOCTEST_CODE_WRAPPER: &str = "fn f(){";
     let wrapped_source = format!("{DOCTEST_CODE_WRAPPER}{source}\n}}");
 
     // Any parse errors should also appear when the doctest is compiled for real,
     // so just suppress them here.
     let emitter = SilentEmitter { translator: rustc_driver::default_translator() };
 
-    // FIXME(misdreavus): pass `-Z treat-err-as-bug` to the doctest parser
     let dcx = DiagCtxt::new(Box::new(emitter)).disable_warnings();
     let psess = ParseSess::with_dcx(dcx, Arc::new(SourceMap::new(FilePathMapping::empty())));
+    // Reset errors at the end so that we don't panic on delayed bugs.
+    let _defer = rustc_data_structures::defer(|| psess.dcx().reset_err_count());
```

Comment:
Instead of having to ensure that we call `reset_err_count()` at every possible function exit, just use `defer` to make it happen automatically which is a lot more robust and legible.